ai/mcp/src/main/java/com/linecorp/armeria/server/ai/mcp/ArmeriaStreamableServerTransportProvider.java

```diff
@@ -0,0 +1,484 @@
+                       .flatMap(McpStreamableServerSession::closeGracefully)
+                       .then();
+        }).then().doOnSuccess(v -> {
+            sessions.clear();
+            if (keepAliveScheduler != null) {
+                keepAliveScheduler.shutdown();
+            }
+        });
+    }
+
+    /**
+     * Returns the {@link HttpService} that defines the transport's HTTP endpoints.
+     * This {@link HttpService} should be registered to Armeria {@link Server}.
+     *
+     * <p>The {@link HttpService} defines one endpoint with three methods:
+     * <ul>
+     *   <li>GET - For the client listening SSE stream</li>
+     *   <li>POST - For receiving client messages</li>
+     *   <li>DELETE - For removing sessions</li>
+     * </ul>
+     */
+    public HttpService httpService() {
+        return httpService;
+    }
+
+    private final class McpStreamableService extends AbstractHttpService {
+
+        @Override
+        public HttpResponse doGet(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+            if (isClosing) {
+                return HttpResponse.of(HttpStatus.SERVICE_UNAVAILABLE, MediaType.PLAIN_TEXT,
+                                       "Server is shutting down");
+            }
+
+            final List<MediaType> acceptHeaders = req.headers().accept();
+            if (!acceptHeaders.contains(MediaType.EVENT_STREAM)) {
+                return HttpResponse.of(HttpStatus.UNSUPPORTED_MEDIA_TYPE, MediaType.PLAIN_TEXT,
+                                       "Accept header must include text/event-stream");
+            }
+
+            final String sessionId = req.headers().get(HttpHeaders.MCP_SESSION_ID, "");
+            if (sessionId.isEmpty()) {
+                return HttpResponse.of(HttpStatus.BAD_REQUEST, MediaType.PLAIN_TEXT,
+                                       "Missing required header: " + HttpHeaders.MCP_SESSION_ID);
+            }
+            // Use a blocking execute since MCP implementations may perform blocking operations.
+            final CompletableFuture<HttpResponse> future =
+                    CompletableFuture.supplyAsync(() -> handleGet(ctx, req, sessionId),
+                                                  ctx.blockingTaskExecutor());
+            return HttpResponse.of(future);
+        }
+
+        private HttpResponse handleGet(ServiceRequestContext ctx, HttpRequest req, String sessionId) {
+            final McpStreamableServerSession session = sessions.get(sessionId);
+            if (session == null) {
+                return HttpResponse.of(HttpStatus.BAD_REQUEST, MediaType.PLAIN_TEXT,
+                                       "No session found for id: " + sessionId);
+            }
+
+            final McpTransportContext mcpContext = contextExtractor.extract(ctx);
+            final String lastEventId = req.headers().get(HttpHeaders.LAST_EVENT_ID, "");
+            if (!lastEventId.isEmpty()) {
```

Comment:
Note) Noted that if `lastEventId` is specified, only replays are sent back (not new events).